filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,281 @@
+}
+
+// A helper to safely set the session field after waiting on any pending
+// operations.
+func (state *kafkaSessionState) setSession(sess sarama.ConsumerGroupSession) {
+	state.mutex.Lock()
+	// Once we claim the mutex we still wait for any pending ACKs to be
+	// sent. (These may well fail if the session is ending, but that's better
+	// than calling a stale pointer.)
+	state.waitGroup.Wait()
+	state.session = sess
+	state.mutex.Unlock()
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (input *kafkaInput) Run() {
+	input.runOnce.Do(func() {
+		// Track errors
+		go func() {
+			for err := range input.consumerGroup.Errors() {
+				input.log.Errorw("Error reading from kafka", "error", err)
+			}
+		}()
+
+		go func() {
+			for {
+				handler := groupHandler{input: input}
+
+				// Sarama uses standard go contexts to control cancellation, so we need
+				// to wrap our input context channel in that interface.
+				err := input.consumerGroup.Consume(
+					doneChannelContext(input.context.Done), input.config.Topics, handler)
+				if err != nil {
+					input.log.Errorw("Kafka consume error", "error", err)
+				}
+
+				// If Consume returned because the context was cancelled, don't resume.
+				select {
+				case <-input.context.Done:
+					return
+				default:
+				}
+			}
+		}()
+	})
+}
+
+// Wait shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Wait() {
+	input.Stop()
+	// Wait for the consumer group to shut down
+	input.consumerGroup.Close()
+}
+
+// Stop shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Stop() {
+	close(input.context.Done)
+}
+
+func arrayForKafkaHeaders(headers []*sarama.RecordHeader) []interface{} {
+	array := []interface{}{}
+	for _, header := range headers {
+		array = append(array, common.MapStr{
+			"key":   header.Key,
+			"value": header.Value,
+		})
+	}
+	return array
+}
+
+// A barebones implementation of context.Context wrapped around the done
+// channels that are more common in the beats codebase.
+// TODO(faec): Generalize this to a common utility in a shared library
+// (https://github.com/elastic/beats/issues/13125).
+type channelCtx <-chan struct{}
+
+func doneChannelContext(ch <-chan struct{}) context.Context {
+	return channelCtx(ch)
+}
+
+func (c channelCtx) Deadline() (deadline time.Time, ok bool) { return }
+func (c channelCtx) Done() <-chan struct{} {
+	return (<-chan struct{})(c)
+}
+func (c channelCtx) Err() error {
+	select {
+	case <-c:
+		return context.Canceled
+	default:
+		return nil
+	}
+}
+func (c channelCtx) Value(key interface{}) interface{} { return nil }
+
+type groupHandler struct {
+	input *kafkaInput
+}
+
+func (h groupHandler) createEvent(
+	sess sarama.ConsumerGroupSession,
+	claim sarama.ConsumerGroupClaim,
+	message *sarama.ConsumerMessage,
+) beat.Event {
+	event := beat.Event{
+		Timestamp: time.Now(),
+	}
+	eventFields := common.MapStr{
+		"message": string(message.Value),
```

Comment:
@urso , this is why i propose to make it an input source instead of just having a kafka input. `message.Value` being a string is a generalization. in many cases Kafka messages can be written in as Protobuf/Avro messages that would need to be deserialized before the event is written out. a user must be able to provide what Decoder to use on Sarama and how to create an event. kindly consider this use case.

cc: @ph